README
NAME
    `IPC::PerlSSH' - execute remote perl code over an SSH link

SYNOPSIS
     use IPC::PerlSSH;

     my $ips = IPC::PerlSSH->new( Host => "over.there" );

     $ips->eval( "use POSIX qw( uname )" );
     my @remote_uname = $ips->eval( "uname()" );

     # We can pass arguments
     $ips->eval( 'open FILE, ">", $_[0]; print FILE $_[1]; close FILE;',
                 "foo.txt", "Hello, world!" );

     # We can pre-compile stored procedures
     $ips->store( "get_file", 'local $/; 
                               open FILE, "<", $_[0];
                               $_ = <FILE>;
                               close FILE;
                               return $_;' );
     foreach my $file ( @files ) {
        my $content = $ips->call( "get_file", $file );
        ...
     }

     # We can use existing libraries for remote stored procedures
     $ips->use_library( "FS", qw( readfile ) );
     foreach my $file ( @files ) {
        my $content = $ips->call( "readfile", $file );
        ...
     }

DESCRIPTION
    This module provides an object class that provides a mechanism to
    execute perl code in a remote instance of perl running on another host,
    communicated via an SSH link or similar connection. Where it differs
    from most other IPC modules is that no special software is required on
    the remote end, other than the ability to run perl. In particular, it is
    not required that the `IPC::PerlSSH' module is installed there. Nor are
    any special administrative rights required; any account that has shell
    access and can execute the perl binary on the remote host can use this
    module.

  Argument Passing
    The arguments to, and return values from, remote code are always
    transferred as lists of strings. This has the following effects on
    various types of values:

    *       String values are passed as they stand.

    *       Booleans and integers will become stringified, but will work as
            expected once they reach the other side of the connection.

    *       Floating-point numbers will get converted to a decimal notation,
            which may lose precision.

    *       A single array of strings, or a single hash of string values,
            can be passed by-value as a list, possibly after positional
            arguments:

             $ips->store( 'foo', 'my ( $arg, @list ) = @_; ...' );

             $ips->store( 'bar', 'my %opts = @_; ...' );

    *       No reference value, including IO handles, can be passed; instead
            it will be stringified.

    To pass or return a more complex structure, consider using a module such
    as Storable, which can serialise the structure into a plain string, to
    be deserialised on the remote end. Be aware however, that `Storable' was
    only added to core in perl 5.7.3, so if the remote perl is older, it may
    not be available.

    To work with remote IO handles, see the IPC::PerlSSH::Library::IO
    module.

CONSTRUCTORS
  $ips = IPC::PerlSSH->new( Host => $host, ... )
    Returns a new instance of a `IPC::PerlSSH' object connected to the
    specified host. The following arguments can be specified:

    Host => STRING
            Connect to a named host.

    Port => INT
            Optionally specify a non-default port.

    Perl => STRING
            Optionally pass in the path to the perl binary in the remote
            host.

    User => STRING
            Optionally pass in an alternative username

    SshPath => STRING
            Optionally specify a different path to the ssh binary

    SshOptions => ARRAY
            Optionally specify any other options to pass to the ssh binary,
            in an `ARRAY' reference

  $ips = IPC::PerlSSH->new( Command => \@command, ... )
    Returns a new instance of a `IPC::PerlSSH' object which uses the
    STDIN/STDOUT streams of a command it executes, as the streams to
    communicate with the remote perl.

    Command => ARRAY
            Specifies the command to execute

    Command => STRING
            Shorthand form for executing a single simple path

    The `Command' key can be used to create an `IPC::PerlSSH' running perl
    directly on the local machine, for example; so that the "remote" perl is
    in fact running locally, but still in its own process.

     my $ips = IPC::PerlSSH->new( Command => $^X );

  $ips = IPC::PerlSSH->new( Readh => $rd, Writeh => $wr )
    Returns a new instance of a `IPC::PerlSSH' object using a given pair of
    filehandles to read from and write to the remote perl process. It is
    allowable for both filehandles to be the same - for example using a
    socket.

  $ips = IPC::PerlSSH->new( Readfunc => \&read, Writefunc => \&write )
    Returns a new instance of a `IPC::PerlSSH' object using a given pair of
    functions as read and write operators.

    Usually this form won't be used in practice; it largely exists to assist
    the test scripts. But since it works, it is included in the interface in
    case the earlier alternatives are not suitable.

    The functions are called as

     $len = $Readfunc->( my $buffer, $maxlen );

     $len = $Writewrite->( $buffer );

    In each case, the returned value should be the number of bytes read or
    written.

METHODS
  @result = $ips->eval( $code, @args )
    This method evaluates code in the remote host, passing arguments and
    returning the result.

    The code should be passed in a string, and is evaluated using a string
    `eval' in the remote host, in list context. If this method is called in
    scalar context, then only the first element of the returned list is
    returned.

    If the remote code threw an exception, then this function propagates it
    as a plain string. If the remote process exits before responding, this
    will be propagated as an exception.

  $ips->store( $name, $code )
  $ips->store( %funcs )
    This method sends code to the remote host to store in named procedure(s)
    which can be executed later. The code should be passed in strings.

    While the code is not executed, it will still be compiled into CODE
    references in the remote host. Any compile errors that occur will be
    throw as exceptions by this method.

    Multiple functions may be passed in a hash, to reduce the number of
    network roundtrips, which may help latency.

  $ips->bind( $name, $code )
    This method is identical to the `store' method, except that the remote
    function will be available as a plain function within the local perl
    program, as a function of the given name in the caller's package.

  @result = $ips->call( $name, @args )
    This method invokes a remote method that has earlier been defined using
    the `store' or `bind' methods. The arguments are passed and the result
    is returned in the same way as with the `eval' method.

    If an exception occurs during execution, it is propagated and thrown by
    this method. If the remote process exits before responding, this will be
    propagated as an exception.

  $ips->use_library( $library, @funcs )
    This method loads a library of code from a module, and stores them to
    the remote perl by calling `store' on each one. The `$library' name may
    be a full class name, or a name within the `IPC::PerlSSH::Library::'
    space.

    If the `@funcs' list is non-empty, then only those named functions are
    stored (analogous to the `use' perl statement). This may be useful in
    large libraries that define many functions, only a few of which are
    actually used.

    For more information, see IPC::PerlSSH::Library.

AUTHOR
    Paul Evans <leonerd@leonerd.org.uk>



Changes
Revision history for IPC-PerlSSH

0.16    CHANGES:
         * Provide a nicer ->build_command_from() method in Base that deletes
           hashref keys
         * Allow arguments or return values to include undef
         * Added a small example script

0.15    CHANGES:
         * Added Library::FS::remove, mkpath, rmtree
         * Added 'Port' and 'SshOptions' constructor arguments
         * Added 'Readh'+'Writeh' constructor form
         * Use Test::Fatal instead of Test::Exception
         * Various documentation updates

0.14    CHANGES:
         * Provide STDERR-capturing versions of Library::Run functions
           (fixes https://rt.cpan.org/Ticket/Display.html?id=49440)

0.13    BUGFIXES:
         * Avoid fchmod/fchown on perls before 5.8.8

0.12    BUGFIXES:
         * Handle SIGPIPE in Run::system_inout, handle exec() failures
         * Always load library initialisation, even with partial load lists
         * Fix tests/libraries to work on perl versions prior to 5.10 -
           package variables need re-declaring

0.11    CHANGES:
         * Added 'use warnings'
         * Created Run and IO libraries
         * Apply 'package' statements to remote stored code and libraries;
           allow libaries to persist package variables
         * Report the offending line of code in remote compile errors
         * Various other code and documentation updates

0.10    CHANGES:
         * Documentation fix to Library::FS::writefile()
         * Neater behaviour in Base.pm for subclasses when parse_message()
           fails

0.09    CHANGES:
         * Neater cleanup behaviour on unref
         * Added 'rename' FS library function

0.08    CHANGES:
         * Added Changes notes file
         * Added ->use_library() functionallity
         * Created FS library
         * Allow alternate path to SSH binary


t/Stash.pm
package t::Stash;

use IPC::PerlSSH::Library;

init 'our %pad;';

func put => '$pad{$_[0]} = $_[1]';

func get => '$pad{$_[0]}';

1;


lib/IPC/PerlSSH/Base.pm
#  You may distribute under the terms of either the GNU General Public License
#  or the Artistic License (the same terms as Perl itself)
#
#  (C) Paul Evans, 2006-2012 -- leonerd@leonerd.org.uk

package IPC::PerlSSH::Base;

use strict;
use warnings;

use Carp;

our $VERSION = '0.16';

=head1 NAME

C<IPC::PerlSSH::Base> - base functionallity behind L<IPC::PerlSSH>

=head1 DESCRIPTION

This module provides the low-level message formatting and parsing code used by
C<IPC::PerlSSH>, and the perl code to be executed on the remote server once a
connection is established.

This split exists, in order to make it easier to write other modules that use
the same behaviour. For example, an asynchronous version could be written
using this as a base class.

=cut

# Some functions we'll share with the remote
my $SHARE_PERL = <<'EOP';
sub pack_message
{
   my ( $message, @args ) = @_;

   my $buffer = "$message\n";
   $buffer .= scalar( @args ) . "\n";

   foreach my $arg ( @args ) {
      if( !defined $arg ) {
         $buffer .= "U\n";
      }
      else {
         $buffer .= length( $arg ) . "\n" . "$arg";
      }
   }

   return $buffer;
}

EOP

eval "$SHARE_PERL; 1" or die $@;

# And now for the main loop of the remote firmware
my $REMOTE_PERL = <<'EOP';
sub send_message
{
   my ( $message, @args ) = @_;
   print STDOUT pack_message( $message, @args );
}

sub read_message
{
   local $/ = "\n";

   my $message = <STDIN>;
   defined $message or return "QUIT";
   chomp $message;

   my $numargs = <STDIN>;
   defined $numargs or die "Expected number of arguments\n";
   chomp $numargs;

   my @args;
   while( $numargs ) {
      my $arglen = <STDIN>;
      defined $arglen or die "Expected length of argument\n";
      chomp $arglen;

      if( $arglen eq "U" ) {
         push @args, undef;
      }
      else {
         my $arg = "";
         while( $arglen ) {
            my $n = read( STDIN, $arg, $arglen, length $arg );
            die "read() returned $!\n" unless( defined $n );
            $arglen -= $n;
         }

         push @args, $arg;
      }
      $numargs--;
   }

   return ( $message, @args );
}

my %compilers;
my %stored_procedures;

sub store
{
   my ( $package, %subs ) = @_;

   my $compiler;
   unless( $compiler = $compilers{$package} ) {
      my $preamble = delete $subs{_init} || "";
      $compiler = eval "package $package; $preamble; sub { eval \$_[0] }";
      if( $@ ) {
         send_message( "DIED", "While compiling initialisation code: $@" );
         return;
      }
      $compilers{$package} = $compiler;
   }

   foreach my $name ( keys %subs ) {
      $stored_procedures{$name} = $compiler->( "sub { $subs{$name} }" ) and next;
      send_message( "DIED", "While compiling code for $name: $@" );
      return;
   }

   send_message( "OK" );
   return;
}

while( 1 ) {
   my ( $message, @args ) = read_message;

   if( $message eq "QUIT" ) {
      # Immediate controlled shutdown
      exit( 0 );
   }

   if( $message eq "EVAL" ) {
      my $code = shift @args;

      my $subref = eval "sub { $code }";
      if( $@ ) {
         send_message( "DIED", "While compiling code: $@" );
         next;
      }

      my @results = eval { $subref->( @args ) };
      if( $@ ) {
         send_message( "DIED", "While running code: $@" );
         next;
      }

      send_message( "RETURNED", @results );
      next;
   }
   
   if( $message eq "STORE" ) {
      store( "main", @args );
      next;
   }

   if( $message eq "STOREPKG" ) {
      store( @args );
      next;
   }

   if( $message eq "CALL" ) {
      my $name = shift @args;

      my $subref = $stored_procedures{$name};
      if( !defined $subref ) {
         send_message( "DIED", "No such stored procedure '$name'" );
         next;
      }

      my @results = eval { $subref->( @args ) };
      if( $@ ) {
         send_message( "DIED", "While running code: $@" );
         next;
      }

      send_message( "RETURNED", @results );
      next;
   }

   send_message( "DIED", "Unknown message $message" );
}
EOP

sub build_command
{
   my $self = shift;
   my %opts = @_;
   return $self->build_command_from( \%opts );
}

sub build_command_from
{
   my $self = shift;
   my ( $opts ) = @_;

   my @command;
   if( exists $opts->{Command} ) {
      my $c = delete $opts->{Command};
      @command = ref($c) && UNIVERSAL::isa( $c, "ARRAY" ) ? @$c : ( "$c" );
   }
   else {
      my $host = delete $opts->{Host} or
         croak ref($self)." requires a Host, a Command or a Readfunc/Writefunc pair";

      defined $opts->{User} and $host = "$opts->{User}\@$host";
      delete $opts->{User};

      my @options;

      push @options, "-p", delete $opts->{Port} if defined $opts->{Port};

      push @options, @{ delete $opts->{SshOptions} } if defined $opts->{SshOptions};

      @command = ( delete $opts->{SshPath} || "ssh", @options, $host, delete $opts->{Perl} || "perl" );
   }

   return @command;
}

sub send_firmware
{
   my $self = shift;

   $self->write( <<EOF );
use strict;
\$| = 1;
$SHARE_PERL

$REMOTE_PERL
__END__
EOF
}

sub parse_message
{
   my $self = shift;
   my $buffer = $_[0]; # We'll assign it back at the end if successful

   $buffer =~ s/^(.*)\n(.*)\n// or return;
   my ( $message, $numargs ) = ( $1, $2 );

   my @args;
   while( $numargs ) {
      $buffer =~ s/^(.*)\n// or return;
      my $arglen = $1;

      if( $arglen eq "U" ) {
         push @args, undef;
      }
      else {
         length $buffer >= $arglen or return;
         my $arg = substr( $buffer, 0, $arglen, "" );
         push @args, $arg;
      }
      $numargs--;
   }

   # If we got this far, we've successfully parsed a message. Reassign the
   # buffer back again
   $_[0] = $buffer;

   return ( $message, @args );
}

# Internal methods
sub write_message
{
   my $self = shift;
   my ( $message, @args ) = @_;

   $self->write( pack_message( $message, @args ) );
}

sub load_library_pkg
{
   my $self = shift;
   my ( $library, @funcs ) = @_;

   require IPC::PerlSSH::Library;

   my $classname;

   # $library may or may not have IPC::PerlSSH::Library:: prefix... try both
   # ways
   foreach my $module ( "IPC::PerlSSH::Library::$library", "$library" ) {
      ( my $filename = $module ) =~ s{::}{/}g; $filename .= ".pm";

      $classname = $module;

      eval { require $filename } and last;

      undef $classname;

      # Examine the error - if we can't find it, go on to the next.
      # Anything else, we'll throw a wobbly. Make sure also it's this file
      # in particular it wants, as opposed to one of its dependencies.
      next if $@ =~ m/^Can't locate \Q$filename\E in \@INC/;

      die $@;
   }

   unless( defined $classname ) {
      croak "Cannot find an IPC::PerlSSH library called $library";
   }

   my %funcs = IPC::PerlSSH::Library::funcs( $classname, @funcs );

   # Don't bother loading again anything already present
   foreach my $name ( keys %funcs ) {
      next if $name =~ m/^_/;

      $self->_has_stored_code( $name ) and delete $funcs{$name};
   }

   return $classname, \%funcs;
}

sub load_library
{
   my $self = shift;
   ( undef, my $funcs ) = $self->load_library_pkg( @_ );
   return %$funcs;
}

# for subclasses to override
sub _has_stored_code
{
   return 0;
}

=head1 AUTHOR

Paul Evans <leonerd@leonerd.org.uk>

=cut

0x55AA;


lib/IPC/PerlSSH/Library.pm
#  You may distribute under the terms of either the GNU General Public License
#  or the Artistic License (the same terms as Perl itself)
#
#  (C) Paul Evans, 2008-2010 -- leonerd@leonerd.org.uk

package IPC::PerlSSH::Library;

use strict;
use warnings;

use Exporter 'import';
our @EXPORT = qw( init func );
use Carp;

our $VERSION = '0.16';

=head1 NAME

C<IPC::PerlSSH::Library> - support package for declaring libraries of remote
functions

=head1 SYNOPSIS

 package IPC::PerlSSH::Library::Info;

 use strict;
 use IPC::PerlSSH::Library;

 func uname   => 'uname()';
 func ostype  => '$^O';
 func perlbin => '$^X';

 1;

This can be loaded by

 use IPC::PerlSSH;

 my $ips = IPC::PerlSSH->new( Host => "over.there" );

 $ips->load_library( "Info" );

 print "Remote perl is running from " . $ips->call("perlbin") . "\n";
 print " Running on a machine of type " . $ips->call("ostype") .
                                          $ips->call("uname") . "\n";

=head1 DESCRIPTION

This module allows the creation of pre-prepared libraries of functions which
may be loaded into a remote perl running via C<IPC::PerlSSH>.

All the code is kept in its own package in the remote perl. The package
declaration is performed in the remote perl, by including an optional block of
initialisation code, passed to the C<init()> function.

Typically this code could C<use> a perl module, or declare shared variables
or functions. Be careful when C<use>ing a module, as the remote perl executing
it may not have the same modules installed as the local machine, or even be of
the same version.

Note that C<our> variables will be available for use in stored code, but
limitations of the way perl's lexical scopes work mean that C<my> variables
will not. On versions of perl before 5.10, the variable will have to be
C<our>ed again in each block of code that requires it. On 5.10 and above, this
is not necessary; but beware that the code will not work on remote perls before
this version, even if the local perl is 5.10.

For example, consider the following small example:

 package IPC::PerlSSH::Library::Storage;

 use IPC::PerlSSH::Library;

 init q{
    our %storage;

    sub list  { keys %storage }
    sub clear { undef %storage }
 };

 func get   => q{ our %storage; return $storage{$_[0]} };
 func set   => q{ our %storage; $storage{$_[0]} = $_[1] };
 func clear => q{ clear() }
 func list  => q{ return list() }

 1;

=cut

my %package_funcs;

=head1 FUNCTIONS

=cut

=head2 func( $name, $code )

Declare a function called $name, which is implemented using the source code in
$code. Note that $code must be a plain string, I<NOT> a CODE reference.

The function name may not begin with an underscore.

=cut

sub func
{
   my ( $name, $code ) = @_;
   my $caller = caller;

   $name =~ m/^_/ and croak "Cannot name a library function beginning with '_'";

   # $code may contain leading whitespace and linefeeds. Kill them
   $code =~ s/\s*\n\s*//g;

   $package_funcs{$caller}->{$name} = $code;
}

=head2 init( $code )

Declare library initialisation code. This code will be executed in the remote
perl before any functions are compiled.

=cut

sub init
{
   my ( $code ) = @_;
   my $caller = caller;

   $package_funcs{$caller}->{_init} and croak "Already have library initialisation";

   # $code may contain leading whitespace and linefeeds. Kill them
   $code =~ s/\s*\n\s*//g;

   $package_funcs{$caller}->{_init} = $code;
}

sub funcs
{
   my ( $classname, @funcs ) = @_;

   my $package_funcs = $package_funcs{$classname};
   $package_funcs or croak "$classname does not define any library functions";

   my %funcs;
   # Always report the _init function
   $funcs{_init} = $package_funcs->{_init} if exists $package_funcs->{_init};

   if( @funcs ) {
      foreach my $f ( @funcs ) {
         $package_funcs->{$f} or croak "$classname does not define a library function called $f";
         $funcs{$f} = $package_funcs->{$f};
      }
   }
   else {
      %funcs = %{ $package_funcs };
   }

   %funcs;
}

=head1 AUTHOR

Paul Evans <leonerd@leonerd.org.uk>

=cut

0x55AA;


lib/IPC/PerlSSH.pm
#  You may distribute under the terms of either the GNU General Public License
#  or the Artistic License (the same terms as Perl itself)
#
#  (C) Paul Evans, 2006-2012 -- leonerd@leonerd.org.uk

package IPC::PerlSSH;

use strict;
use warnings;

use base qw( IPC::PerlSSH::Base );

use IPC::Open2;

use Carp;

our $VERSION = '0.16';

=head1 NAME

C<IPC::PerlSSH> - execute remote perl code over an SSH link

=head1 SYNOPSIS

 use IPC::PerlSSH;

 my $ips = IPC::PerlSSH->new( Host => "over.there" );

 $ips->eval( "use POSIX qw( uname )" );
 my @remote_uname = $ips->eval( "uname()" );

 # We can pass arguments
 $ips->eval( 'open FILE, ">", $_[0]; print FILE $_[1]; close FILE;',
             "foo.txt", "Hello, world!" );

 # We can pre-compile stored procedures
 $ips->store( "get_file", 'local $/; 
                           open FILE, "<", $_[0];
                           $_ = <FILE>;
                           close FILE;
                           return $_;' );
 foreach my $file ( @files ) {
    my $content = $ips->call( "get_file", $file );
    ...
 }

 # We can use existing libraries for remote stored procedures
 $ips->use_library( "FS", qw( readfile ) );
 foreach my $file ( @files ) {
    my $content = $ips->call( "readfile", $file );
    ...
 }

=head1 DESCRIPTION

This module provides an object class that provides a mechanism to execute perl
code in a remote instance of perl running on another host, communicated via an
SSH link or similar connection. Where it differs from most other IPC modules
is that no special software is required on the remote end, other than the
ability to run perl. In particular, it is not required that the
C<IPC::PerlSSH> module is installed there. Nor are any special administrative
rights required; any account that has shell access and can execute the perl
binary on the remote host can use this module.

=head2 Argument Passing

The arguments to, and return values from, remote code are always transferred
as lists of strings. This has the following effects on various types of
values:

=over 8

=item *

String values are passed as they stand.

=item *

Booleans and integers will become stringified, but will work as expected once
they reach the other side of the connection.

=item *

Floating-point numbers will get converted to a decimal notation, which may
lose precision.

=item *

A single array of strings, or a single hash of string values, can be passed
by-value as a list, possibly after positional arguments:

 $ips->store( 'foo', 'my ( $arg, @list ) = @_; ...' );

 $ips->store( 'bar', 'my %opts = @_; ...' );

=item *

No reference value, including IO handles, can be passed; instead it will be
stringified.

=back

To pass or return a more complex structure, consider using a module such as
L<Storable>, which can serialise the structure into a plain string, to be
deserialised on the remote end. Be aware however, that C<Storable> was only
added to core in perl 5.7.3, so if the remote perl is older, it may not be
available.

To work with remote IO handles, see the L<IPC::PerlSSH::Library::IO> module.

=cut

=head1 CONSTRUCTORS

=cut

=head2 $ips = IPC::PerlSSH->new( Host => $host, ... )

Returns a new instance of a C<IPC::PerlSSH> object connected to the specified
host. The following arguments can be specified:

=over 8

=item Host => STRING

Connect to a named host.

=item Port => INT

Optionally specify a non-default port.

=item Perl => STRING

Optionally pass in the path to the perl binary in the remote host.

=item User => STRING

Optionally pass in an alternative username

=item SshPath => STRING

Optionally specify a different path to the F<ssh> binary

=item SshOptions => ARRAY

Optionally specify any other options to pass to the F<ssh> binary, in an
C<ARRAY> reference

=back

=head2 $ips = IPC::PerlSSH->new( Command => \@command, ... )

Returns a new instance of a C<IPC::PerlSSH> object which uses the STDIN/STDOUT
streams of a command it executes, as the streams to communicate with the
remote F<perl>.

=over 8

=item Command => ARRAY

Specifies the command to execute

=item Command => STRING

Shorthand form for executing a single simple path

=back

The C<Command> key can be used to create an C<IPC::PerlSSH> running perl
directly on the local machine, for example; so that the "remote" perl is in
fact running locally, but still in its own process.

 my $ips = IPC::PerlSSH->new( Command => $^X );

=head2 $ips = IPC::PerlSSH->new( Readh => $rd, Writeh => $wr )

Returns a new instance of a C<IPC::PerlSSH> object using a given pair of
filehandles to read from and write to the remote F<perl> process. It is
allowable for both filehandles to be the same - for example using a socket.

=head2 $ips = IPC::PerlSSH->new( Readfunc => \&read, Writefunc => \&write )

Returns a new instance of a C<IPC::PerlSSH> object using a given pair of
functions as read and write operators.

Usually this form won't be used in practice; it largely exists to assist the
test scripts. But since it works, it is included in the interface in case the
earlier alternatives are not suitable.

The functions are called as

 $len = $Readfunc->( my $buffer, $maxlen );

 $len = $Writewrite->( $buffer );

In each case, the returned value should be the number of bytes read or
written.

=cut

sub new
{
   my $class = shift;
   my %opts = @_;

   my $self = bless {
      readbuff => "",
      stored => {},
   }, $class;

   my ( $readfunc, $writefunc ) = ( delete $opts{Readfunc}, delete $opts{Writefunc} );

   my $pid = delete $opts{Pid};

   if( !defined $readfunc || !defined $writefunc ) {
      my ( $readh, $writeh ) = ( delete $opts{Readh}, delete $opts{Writeh} );

      if( !defined $readh || !defined $writeh ) {
         my @command = $self->build_command_from( \%opts );
         $pid = open2( $readh, $writeh, @command );
      }

      $readfunc = sub {
         sysread( $readh, $_[0], $_[1] );
      };

      $writefunc = sub {
         syswrite( $writeh, $_[0] );
      };
   }

   keys %opts and
      croak "Unexpected ->new keys - " . join ", ", sort keys %opts;

   $self->{pid}       = $pid;
   $self->{readfunc}  = $readfunc;
   $self->{writefunc} = $writefunc;

   $self->send_firmware;

   return $self;
}

sub write
{
   my $self = shift;
   my ( $data ) = @_;

   $self->{writefunc}->( $data );
}

sub read_message
{
   my $self = shift;

   my ( $message, @args );

   while( !defined $message ) {
      my $b;
      $self->{readfunc}->( $b, 8192 ) or return ( "CLOSED" );
      $self->{readbuff} .= $b;
      ( $message, @args ) = $self->parse_message( $self->{readbuff} );
   }

   return ( $message, @args );
}

=head1 METHODS

=cut

=head2 @result = $ips->eval( $code, @args )

This method evaluates code in the remote host, passing arguments and returning
the result.

The code should be passed in a string, and is evaluated using a string
C<eval> in the remote host, in list context. If this method is called in
scalar context, then only the first element of the returned list is returned.

If the remote code threw an exception, then this function propagates it as a
plain string. If the remote process exits before responding, this will be
propagated as an exception.

=cut

sub eval
{
   my $self = shift;
   my ( $code, @args ) = @_;

   $self->write_message( "EVAL", $code, @args );

   my ( $ret, @retargs ) = $self->read_message;

   if( $ret eq "RETURNED" ) {
      # If the caller didn't want an array and we received more than one result
      # from the far end; we'll just have to throw it away...
      return wantarray ? @retargs : $retargs[0];
   }
   elsif( $ret eq "DIED" ) {
      my ( $message ) = @retargs;
      if( $message =~ m/^While compiling code:.* at \(eval \d+\) line (\d+)/ ) {
         $message .= " ==> " . (split m/\n/, $code)[$1 - 1] . "\n";
      }
      die "Remote host threw an exception:\n$message";
   }
   elsif( $ret eq "CLOSED" ) {
      die "Remote connection closed\n";
   }
   else {
      die "Unknown return result $ret\n";
   }
}

=head2 $ips->store( $name, $code )

=head2 $ips->store( %funcs )

This method sends code to the remote host to store in named procedure(s) which
can be executed later. The code should be passed in strings.

While the code is not executed, it will still be compiled into CODE references
in the remote host. Any compile errors that occur will be throw as exceptions
by this method.

Multiple functions may be passed in a hash, to reduce the number of network
roundtrips, which may help latency.

=cut

sub store
{
   my $self = shift;
   my %funcs = @_;

   foreach my $name ( keys %funcs ) {
      $self->_has_stored_code( $name ) and croak "Already have a stored function called '$name'";
   }

   $self->write_message( "STORE", %funcs );

   my ( $ret, @retargs ) = $self->read_message;

   if( $ret eq "OK" ) {
      $self->{stored}{$_} = 1 for keys %funcs;
      return;
   }
   elsif( $ret eq "DIED" ) {
      my ( $message ) = @retargs;
      if( $message =~ m/^While compiling code for (\S+):.* at \(eval \d+\) line (\d+)/ ) {
         my $code = $funcs{$1};
         $message .= " ==> " . (split m/\n/, $code)[$2 - 1] . "\n";
      }
      die "Remote host threw an exception:\n$message";
   }
   elsif( $ret eq "CLOSED" ) {
      die "Remote connection closed\n";
   }
   else {
      die "Unknown return result $ret\n";
   }
}

sub _has_stored_code
{
   my $self = shift;
   my ( $name ) = @_;
   return exists $self->{stored}{$name};
}

=head2 $ips->bind( $name, $code )

This method is identical to the C<store> method, except that the remote
function will be available as a plain function within the local perl
program, as a function of the given name in the caller's package.

=cut

sub bind
{
   my $self = shift;
   my ( $name, $code ) = @_;

   $self->store( $name, $code );

   my $caller = (caller)[0];
   {
      no strict 'refs';
      *{$caller."::$name"} = sub { $self->call( $name, @_ ) };
   }
}

=head2 @result = $ips->call( $name, @args )

This method invokes a remote method that has earlier been defined using the
C<store> or C<bind> methods. The arguments are passed and the result is
returned in the same way as with the C<eval> method.

If an exception occurs during execution, it is propagated and thrown by this
method. If the remote process exits before responding, this will be propagated
as an exception.

=cut

sub call
{
   my $self = shift;
   my ( $name, @args ) = @_;

   $self->_has_stored_code( $name ) or croak "Do not have a stored function called '$name'";

   $self->write_message( "CALL", $name, @args );

   my ( $ret, @retargs ) = $self->read_message;

   if( $ret eq "RETURNED" ) {
      # If the caller didn't want an array and we received more than one result
      # from the far end; we'll just have to throw it away...
      return wantarray ? @retargs : $retargs[0];
   }
   elsif( $ret eq "DIED" ) {
      die "Remote host threw an exception:\n$retargs[0]";
   }
   elsif( $ret eq "CLOSED" ) {
      die "Remote connection closed\n";
   }
   else {
      die "Unknown return result $ret\n";
   }
}

=head2 $ips->use_library( $library, @funcs )

This method loads a library of code from a module, and stores them to the
remote perl by calling C<store> on each one. The C<$library> name may be a
full class name, or a name within the C<IPC::PerlSSH::Library::> space.

If the C<@funcs> list is non-empty, then only those named functions are stored
(analogous to the C<use> perl statement). This may be useful in large
libraries that define many functions, only a few of which are actually used.

For more information, see L<IPC::PerlSSH::Library>.

=cut

sub use_library
{
   my $self = shift;

   my ( $package, $funcs ) = $self->load_library_pkg( @_ );

   $self->{stored_pkg}{$package} and delete $funcs->{_init};

   $self->write_message( "STOREPKG", $package, %$funcs );

   my ( $ret, @retargs ) = $self->read_message;

   if( $ret eq "OK" ) {
      $self->{stored_pkg}{$package} = 1;
      $self->{stored}{$_} = 1 for keys %$funcs;
      return;
   }
   elsif( $ret eq "DIED" ) {
      die "Remote host threw an exception:\n$retargs[0]";
   }
   elsif( $ret eq "CLOSED" ) {
      die "Remote connection closed\n";
   }
   else {
      die "Unknown return result $ret\n";
   }
}

sub DESTROY
{
   my $self = shift;

   undef $self->{readfunc};
   undef $self->{writefunc};
   # This will clean up the closures, and hence close the filehandles that are
   # referenced by them. The remote perl will then shut down, and we can wait
   # for the child process to exit

   waitpid $self->{pid}, 0 if defined $self->{pid};
}

=head1 AUTHOR

Paul Evans <leonerd@leonerd.org.uk>

=cut

0x55AA;
